src/buffer/buffer.rs

```diff
@@ -186,65 +181,65 @@ impl Buffer {
     }
 
     /// Print a string, starting at the position (x, y)
-    ///
-    /// `style` accepts any type that is convertible to [`Style`] (e.g. [`Style`], [`Color`], or
-    /// your own type that implements [`Into<Style>`]).
     pub fn set_string<T, S>(&mut self, x: u16, y: u16, string: T, style: S)
     where
         T: AsRef<str>,
         S: Into<Style>,
     {
-        self.set_stringn(x, y, string, usize::MAX, style.into());
+        self.set_stringn(x, y, string, usize::MAX, style);
     }
 
     /// Print at most the first n characters of a string if enough space is available
-    /// until the end of the line
+    /// until the end of the line.
     ///
-    /// `style` accepts any type that is convertible to [`Style`] (e.g. [`Style`], [`Color`], or
-    /// your own type that implements [`Into<Style>`]).
+    /// Use [`Buffer::set_string`] when the maximum amount of characters can be printed.
     pub fn set_stringn<T, S>(
         &mut self,
         x: u16,
         y: u16,
         string: T,
-        width: usize,
+        max_width: usize,
         style: S,
     ) -> (u16, u16)
     where
         T: AsRef<str>,
         S: Into<Style>,
     {
-        let style = style.into();
         let mut index = self.index_of(x, y);
-        let mut x_offset = x as usize;
-        let graphemes = UnicodeSegmentation::graphemes(string.as_ref(), true);
-        let max_offset = min(self.area.right() as usize, width.saturating_add(x as usize));
-        for s in graphemes {
-            let width = s.width();
-            if width == 0 {
-                continue;
-            }
-            // `x_offset + width > max_offset` could be integer overflow on 32-bit machines if we
-            // change dimensions to usize or u32 and someone resizes the terminal to 1x2^32.
-            if width > max_offset.saturating_sub(x_offset) {
-                break;
-            }
-
-            self.content[index].set_symbol(s);
-            self.content[index].set_style(style);
+        let max_width = max_width.try_into().unwrap_or(u16::MAX);
+        let mut remaining_width = self.area.right().saturating_sub(x).min(max_width);
+        let graphemes = UnicodeSegmentation::graphemes(string.as_ref(), true)
+            .map(|symbol| (symbol, symbol.width() as u16))
+            .filter(|(_symbol, width)| *width > 0)
+            .map_while(|(symbol, width)| {
+                remaining_width = remaining_width.checked_sub(width)?;
+                Some((symbol, width))
+            });
+        let style = style.into();
+        let mut x_offset = x;
+        for (symbol, width) in graphemes {
+            self.content
+                .get_mut(index)
+                .expect("Index only exists when inside the buffer area")
```

Comment:
This message seems a bit vague to me. From the caller of this method's perspective there is no index parameter, so this would make no sense to the user.

Reply:
```suggestion
                .unwrap() // Index only exists when inside the buffer area
```

Reply:
This goes away with the simplification, and you get errors about (x, y) instead coming from the get_mut(x, y).
I had an error in the simplification when I first wrote it, where the `x+=1` came before the `next_symbol = x+width` which highlighted this.

Reply:
"goes away" → silently panic via `debug_assert` now :ghost:

Reply:
x/y should be correct for the first call (this is the same behavior of set_stringn before this) and following calls are ensured to be correct by remaining width. The `expect` was mainly a "this should never fail when the above code is correct" message.

In the long run #1011 is a good idea, but this method is unlikely to be problematic after the first call. And as no one seems to care about this panicing on bad x/y input it's not that important to fix this quickly.